x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -618,20 +551,234 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
+      const redactedSpacesWithMatchingAliases =
+        spacesWithMatchingAliases &&
+        getRedactedSpaces(type, 'bulk_get', typeActionMap, spacesWithMatchingAliases);
+      return {
+        ...obj,
+        spaces: redactedSpaces,
+        ...(redactedSpacesWithMatchingAliases && {
+          spacesWithMatchingAliases: redactedSpacesWithMatchingAliases,
+        }),
+        inboundReferences: redactedInboundReferences,
+      };
+    });
+
+    return {
+      objects: filteredAndRedactedObjects,
+    };
+  }
+
+  public async updateObjectsSpaces(
+    objects: SavedObjectsUpdateObjectsSpacesObject[],
+    spacesToAdd: string[],
+    spacesToRemove: string[],
+    options: SavedObjectsUpdateObjectsSpacesOptions = {}
+  ) {
+    const { namespace } = options;
+    const currentSpaceId = SavedObjectsUtils.namespaceIdToString(namespace); // We need this whether the Spaces plugin is enabled or not.
+
+    const allSpacesSet = new Set<string>([currentSpaceId, ...spacesToAdd, ...spacesToRemove]);
+    const bulkGetResponse = await this.baseClient.bulkGet(objects, { namespace });
+    const objectsToUpdate = objects.map(({ type, id }, i) => {
+      const { namespaces: spaces = [], version } = bulkGetResponse.saved_objects[i];
+      // If 'namespaces' is undefined, the object was not found (or it is namespace-agnostic).
+      // Either way, we will pass in an empty 'spaces' array to the base client, which will cause it to skip this object.
+      for (const space of spaces) {
+        if (space !== ALL_SPACES_ID) {
+          // If this is a specific space, add it to the spaces we'll check privileges for (don't accidentally check for global privileges)
+          allSpacesSet.add(space);
+        }
+      }
+      return { type, id, spaces, version };
+    });
+
+    const uniqueTypes = this.getUniqueObjectTypes(objects);
+    const { typeActionMap } = await this.ensureAuthorized(
+      uniqueTypes,
+      ['bulk_get', 'share_to_space'],
+      Array.from(allSpacesSet),
+      { requireFullAuthorization: false }
+    );
+
+    const addToSpaces = spacesToAdd.length ? spacesToAdd : undefined;
+    const deleteFromSpaces = spacesToRemove.length ? spacesToRemove : undefined;
+    try {
+      // The user must be authorized to share every requested object in each of: the current space, spacesToAdd, and spacesToRemove.
+      const spaces = this.getUniqueSpaces(currentSpaceId, ...spacesToAdd, ...spacesToRemove);
+      this.ensureAuthorizedInAllSpaces(objects, 'share_to_space', typeActionMap, spaces);
+    } catch (error) {
+      objects.forEach(({ type, id }) =>
+        this.auditLogger.log(
+          savedObjectEvent({
+            action: SavedObjectAction.UPDATE_OBJECTS_SPACES,
+            savedObject: { type, id },
+            addToSpaces,
+            deleteFromSpaces,
+            error,
+          })
+        )
+      );
+      throw error;
+    }
+    for (const { type, id } of objectsToUpdate) {
+      this.auditLogger.log(
```

Comment:
Well, we skip audit records for bulkGet because those are access events. But these are update events -- we aren't recording what is actually updated, we are recording what the user is attempting to update. So I think a better analogy here is bulkUpdate, and we do add audit records for objects that potentially don't exist. So this seemed like the most consistent thing to do here, even though it's not ideal.

Reply:
I'll buy that argument!